pkg/expression/expression.go

```diff
@@ -1171,26 +1171,37 @@ func wrapWithIsTrue(ctx BuildContext, keepNull bool, arg Expression, wrapForInt
 		}
 	}
 	var fc *isTrueOrFalseFunctionClass
+	var fn ast.CIStr
 	if keepNull {
 		fc = &isTrueOrFalseFunctionClass{baseFunctionClass{ast.IsTruthWithNull, 1, 1}, opcode.IsTruth, keepNull}
+		fn = ast.NewCIStr(ast.IsTruthWithNull)
 	} else {
 		fc = &isTrueOrFalseFunctionClass{baseFunctionClass{ast.IsTruthWithoutNull, 1, 1}, opcode.IsTruth, keepNull}
+		fn = ast.NewCIStr(ast.IsTruthWithoutNull)
 	}
-	f, err := fc.getFunction(ctx, []Expression{arg})
+	f, err := fc.getFunction(ctx, []Expression{simpleExpression(arg)})
 	if err != nil {
 		return nil, err
 	}
 	sf := &ScalarFunction{
-		FuncName: ast.NewCIStr(ast.IsTruthWithoutNull),
+		FuncName: fn,
 		Function: f,
 		RetType:  f.getRetTp(),
 	}
-	if keepNull {
-		sf.FuncName = ast.NewCIStr(ast.IsTruthWithNull)
-	}
 	return FoldConstant(ctx, sf), nil
 }
 
+func simpleExpression(arg Expression) Expression {
```

Comment:
The convert is too aggressive, istrue(cast(xxx as xxx)) is not always equal to istrue(xxx)
For example:
```

mysql> select istrue_with_null('0.1');
+-------------------------+
| istrue_with_null('0.1') |
+-------------------------+
|                       1 |
+-------------------------+
1 row in set (0.01 sec)

mysql> select istrue_with_null(cast('0.1' as signed));
+-----------------------------------------+
| istrue_with_null(cast('0.1' as signed)) |
+-----------------------------------------+
|                                       0 |
+-----------------------------------------+
1 row in set, 1 warning (0.00 sec)

```